### ISSUE TITLE
Migrations and deployment experience

### ISSUE THREAD

**Issue by ajcvickers** [dotnet/efcore#19587]
This an "epic" issue for the theme of improving the Migrations and deployment experience. Specific pieces of work will be tracked by linked issues.

----

Currently, many developers migrate their databases at application startup time. This is easy but is not recommended because:

* Multiple threads/processes/servers may attempt to migrate the database concurrently
* Applications may try to access inconsistent state while this is happening
* Usually the database permissions to modify the schema should not be granted for application execution
* Its hard to revert back to a clean state if something goes wrong

We want to deliver a better experience here that allows an easy way to migrate the database at deployment time. This should:

* Work on Linux, Mac, and Windows
* Be a good experience on the command line
* Support scenarios with containers
* Work with commonly used real-world deployment tools/flows
* Integrate into at least Visual Studio

The result is likely to be many small improvements in EF Core (for example, better Migrations on SQLite), together with guidance and longer-term collaborations with other teams to improve end-to-end experiences that go beyond just EF.

**Comment by ErikEJ** [comment]
sqlpackage / dacpac already does much of this - but sadly currently only for SQL Server.

**Comment by BEagle1984** [comment]
In our latest project we integrated the DB migrations in the deployment pipeline simply by generating an idempotent migration script (for all migration steps) via dotnet-ef and leveraging sqlcmd to execute it against the appropriate sql server instance. We do this in Jenkins, from a Linux slave.
I honestly don’t think our solution is too bad.

**Comment by bricelam** [comment]
Some Related issues:
https://github.com/dotnet/efcore/issues/7681
https://github.com/dotnet/efcore/issues/8695
https://github.com/dotnet/efcore/issues/9841
https://github.com/dotnet/efcore/issues/10750
https://github.com/dotnet/efcore/issues/12911
https://github.com/dotnet/efcore/issues/14349
https://github.com/dotnet/efcore/issues/14531
https://github.com/dotnet/efcore/issues/14746
https://github.com/dotnet/efcore/issues/15174
https://github.com/dotnet/efcore/issues/16175
https://github.com/dotnet/efcore/issues/16406
https://github.com/dotnet/efcore/issues/18624
https://github.com/dotnet/efcore/issues/18840
(TODO: File an issue for Migrations "bundles")
(TODO: File an issue for Migrations projects)

**Comment by macsux** [comment]
So I've built something that helps with this that was contributed to project SteelToe. It allows tasks to be bundled with app and launched via special command argument. Ef migration is one of them. 
Is you're not familiar with SteelToe, the site is https://steeltoe.io

These are the relevant sources you may consider

https://github.com/SteeltoeOSS/Management
https://github.com/SteeltoeOSS/Connectors/blob/dev/src/Steeltoe.CloudFoundry.Connector.EFCore/MigrateDbContextTask.cs

**Comment by johnnyreilly** [comment]
Thanks for linking to this @AndriySvyryd - great to see this is being actively worked on

**Comment by dazinator** [comment]
I ended up architecting my app to address these points, and putting all the logic in a reusable library which I might open source I  future.

- A lock file in shared storage directory (directory seen by all instances) to ensure only one instance can run the migrations at a time.
- An mvc Action filter which rejects requests  by default when there are EF pending upgrades detected and includes a header in the response so that the client can detect that the server requires ef core migrations to be applied and can display a suitable UI. 
- A blazor component and http client delegate handler to be used in blazor applications that can auto detect the header in responses from the server and display an Upgrade pending screen, with an option for a deployment admin to login and apply. 
- An MVC controller with an Action method that can be called to Apply the pendong EF core migrations, which is excluded from the action filter mentioned above that rejects requests. I've made this method require a deployment admin to be logged in to call it.
- A custom authentication scheme that allows a "deployment admin" to login which is a login method that places no dependency on the application database for obvious reasons. This custom auth scheme generates a code and writes is to a file in the storage directory. A deployment admin can then grab the file from the server / container storage directory and use that code to log in with which creates a claims principal with Deployment Admin role, and they can then click a button to apply the migration (I.e this calls the mvc action method mentioned above)
 
There was quite a bit to it in the end.
I havent addressed rollback.. but being able to rollback is not necessarily safe if data is going to be lost in the process, so the safest policy for me tends to be to only allow roll forward, and make sure migrations are tested adequately before rolling to a production environment.

**Comment by ajcvickers** [comment]
@dazinator Thanks for sharing that. There is certainly some interesting things there--we talked about going in this kind of direction before, but robustness has always been a concern. If you do create a package I would be interested in taking a look.

**Comment by syndicatedshannon** [comment]
This product is pretty strong with this latest 3.0 release. It's great to see, since we decided not to bail on it during the rewrite.

I'm disheartened, though, that this capability is slated for 5.0, but still very little attention is being given to contract-first, conceptual model-first, database-first, or SSDT/.SQLPROJ-authoritative development. Is it just that there are insufficient requests for it, or is EF not intended for that market?

The semantics of up/down are great for early development, and I appreciate that they pull a lot of teams into EF during envisioning stages. Also, the syntax of EF's C# attribute-modelled and fluent definition are well-designed. Even so, this leaves a lot of gaps in both relational and constraint capability and support for collaboration with data designers. SQL code embedded as text strings in fluent definitions, to me, means something is wrong. Which team writes that code? The team I have to allocate with both expert level C# lambda AND T-SQL background?

I know this all reveals my personal stance on code-first data development. The point is, I wonder where this feature is heading. Who needs to do this; but wouldn't be better suited to getting SQL schema definitions out of C# and into source-controlled SQL, or a contract-first toolset?

Thanks for listening.

**Comment by ErikEJ** [comment]
@syndicatedshannon I assume you are aware that EF Core Power Tools supports exactly that workflow (database first from a .dacpac/ .sqlpoj)

**Comment by syndicatedshannon** [comment]
@ErikEJ - Thank you. Our team has visited that tool often in the past, starting when you first suggested it years ago. We had various challenges along the way. At last visit, I believe we did not see how to make it work for us to get schema revisions applied to our models as part of our toolchain. My recollection is there is an effortful process to replace the model each time, even if we weren't concerned about losing model customizations. Since then we purchased LLBLGen, but I would love to know if I'm missing something.

**Comment by syndicatedshannon** [comment]
@ErikEJ If you think I'm mistaken, and it does indeed support exactly this workflow, please LMK. Would love to hear I've missed something.

**Comment by ErikEJ** [comment]
@syndicatedshannon The workflow you describe works fine for me with EF Core Power Tools. You can extend the generated POCO classes with NonMapped properties, as these classes are partial, and you can modify the DbContext model by implementing the partial OnModelCreatingPartial method in a partial DbContext class. But of course if you start treating your DAL as a Domain model, things could start falling apart. Please let me know what is blocking you from adpoting this, or if anything I wrote is unclear :-)

**Comment by syndicatedshannon** [comment]
Thank you Erik, I will try to understand, and follow up on #4321 when I have a handle on it, referencing your name. Apologize to others for cluttering this topic.

**Comment by GeeWee** [comment]
Just want to chime in on this, as I've been researching this a bit.
It seems like most of the stuff re: parallel migrations can be solved by taking out exclusive locks.

I can see rails does it with [advisory locks for postgres and mysql](https://github.com/rails/rails/pull/22103), [FluentMigrator](https://github.com/fluentmigrator/documentation/pull/11/files) and [Flyway](https://github.com/flyway/flyway/blob/master/flyway-core/src/main/java/org/flywaydb/core/internal/database/sqlserver/SQLServerApplicationLockTemplate.java) does it with `sp_getapplock` for SQL Server, etc. 

It seems that using the database locking can mitigate running migrations in parallel, thus making it safe to run on Startup. I'm not sure if this  is true for *all* databases, but it seems to be for the big ones.

However, this of course only deals with half of the issue, because e.g. for rolling updates there's still an issue if some apps expect a different schema. Of course you can get around this by planning your [migrations and releases well.](https://www.thereformedprogrammer.net/handling-entity-framework-core-database-migrations-in-production-part-2/#applying-breaking-changes-to-a-continuous-running-application-the-five-stage-app-update)

**Comment by ondrej-marinak** [comment]
### Feedback regarding migrations and deployment experience

Setup: 
- Multi-stage CI/CD pipeline
- Build agent on Ubuntu 18.04
- DB server: MariaDB
- multitenancy in the near future (db per tenant) 

Recently, I spent a lot of time trying to solve the migrations and deployment case in the Azure Pipelines. I’ve tried to solve a straight forward case i.e. to update a database during the **deployment stage**.

1. My first idea was to split the whole into two stages, **build** and **deploy**. 
   - Build stage: generate migration script using` dotnet ef migrations script --output script.sql --idempotent `
   - Deployment stage should download the artifact (sql script) and execute it. 

   Problem: It’s not possible to execute the script using `dotnet ef` or using some extensions. The extensions that I tried didn't work. There are probably some extension but are working well only with MS SQL. The solution could be to install db specific executor on the agent machine.

2. The second approach I tried is to migrate directly from assemblies
   - Build stage: nothing to do here
   - Deployment stage: execute `dotnet exec --depsfile [path] --runtimeconfig [path] [ef path] etc.`

   Problems:
   - Command to execute is not that simple
   - Need to know the path to `ef.dll`
   - Error saying that `Microsoft.EntityFrameworkCore.Design` nuget package is not found even if it was referenced in the startup project explicitly

   I also abandoned this approach because I was constantly stumbling upon new hurdles.

3. The third approach, unfortunately, needs to **pull and build source code once again** in the deployment stage, on the other hand, it is just `dotnet ef database update` command. I stuck with this approach for now since it's the easiest one and doesn't require a lot of plumbing, I don't like it, though.

**Comment by JonPSmith** [comment]
Can I add one issue I have had when updating a database with migrations that might be added in a CI/CD solution.

With a client an update to EF Core 3 required some lambda properties to be added to the database. The classic example is adding a `FullName` string property to replace the lambda property that created the `FullName` from the `FirstName` and `LastName`.

The migration was easy, but updating the thousands of rows to have the correct `FullName` was a problem. My client's system used 'migrate on startup' and I added code to run code when the specific migration was applied. This code then filled in the new `FullName` from the `FirstName` and `LastName` columns. The problem was Azure WebApp timed out because the ASP.NET Core app took too long to start.

If the CI/CD migration could call code to run after the migration and provide names of migration just applied to the database, then this would handle this sort of issue.

**Comment by joaopgrassi** [comment]
> The migration was easy, but updating the thousands of rows to have the correct FullName was a problem. My client's system used 'migrate on startup' and I added code to run code when the specific migration was applied. This code then filled in the new FullName from the FirstName and LastName columns. The problem was Azure WebApp timed out because the ASP.NET Core app took too long to start.

I think doing this at startup blocking is not very good.. maybe you could create a HostedService and do it in the background so the main app is not blocked. Not sure this is a migration or even CI/CD problem.

**Comment by JonPSmith** [comment]
Hi @joaopgrassi,

I did consider using a `BackgroundService`, but that would mean that the application would start with inaccurate data. I deemed that was not a good idea.

The other option was `IHostedService`, which runs before the main ASP.NET Core is run (since version 3.0). But I thought that Azure would time out because it wasn't responding to HTTP requests (I didn't try that - its pretty hard to try these things as you have to reset the database every time).

**Comment by PieterjanDeClippel** [comment]
https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-5.0/plan#migrations-and-deployment-experience explains why it's not a good idea to use `DbContext.Database.EnsureCreated()` at runtime. The recommended approach is to generate a database script using the following command:

    dotnet ef migrations script [FromMigrationExclusive] --output "MigrationScripts\[ScriptName].sql" --context [YourDbContext]

And run this script manually with a user that has permissions to change the database structure.

> - Multiple threads/processes/servers may attempt to migrate the database concurrently
> - Applications may try to access inconsistent state while this is happening _(this is the case when your server is using load balancing)_
> - Usually the database permissions to modify the schema should not be granted for application execution
> - It's hard to revert back to a clean state if something goes wrong

**Comment by vzakanj** [comment]
@ondrej-marinak we are doing it using the first approach you've listed, only with PostgreSQL.
It boils down to installing PostgreSQL command line tools on the agent and caching it with the Azure DevOps cache task so we don't have to install every time the build runs.

1. use psql to query the _EFMigrationsHistory in the db to get the last migration applied for the environment
2. generate idempotent script from the last applied migration to the latest one
3. after all builds are completed, apply the script with psql again

All of the above are done with PowerShell scripts, but I'd like to see this built into dotnet ef tooling. e.g. 
1. not having to manually install psql and query _EfMigrationsHistory + pass the migration as a parameter to dotnet ef migrations script 
2. possibly running the generated script with a separate ef command

**Comment by nick5454** [comment]
@ondrej-marinak I create the ef script in the build pipeline(Cake.DotNetCoreEf) and use the Sql Deploy task in the release and it works fine. The issue I have that if my file has a lot of schema changes it completely fails. Looking to see if I can create a DacPac from a migration after building

**Comment by ajcvickers** [comment]
Unfortunately, while we have made good progress in understanding the space here, we are scoping this issue for the 5.0 release. The migrations bundles feature (#19693) will be deferred until after the EF Core 5.0 release. However, several other targeted improvements related to migrations will be included in EF Core 5.0, including:

* SQLite table rebuilds (#329) (Done)
* Better migrations transaction support (#7681)
* Pass the database connection string on the command line (#10750) (Done)
* Better idempotent scripts for SQL Server (#12911) and other providers (#14746)

**Comment by JonPSmith** [comment]
Can I ask if this might get into a later EF Core 5 release, say 5.1, or will it move to EF Core 6. I only ask this question to know whether I might include information about this in the book that I am updating. 

An answer of "don't know" is fine, and in that case I won't mention it in the EF Core 5 update.

**Comment by ajcvickers** [comment]
@JonPSmith Currently there are no plans for a release between EF Core 5.0 and EF Core 6.0. We're aligned to the yearly cadence of .NET.

**Comment by JonPSmith** [comment]
Thanks. 

Great work on EF Core 5. I will look forward to this feature in the future.

**Comment by alrz** [comment]
I didn't find this mentioned but it would be nice to also improve version control experience.

Currently if more than one dev add migrations the second dev usually needs to redo changes.

**Comment by Saibamen** [comment]
@alrz: +1 for this

**Comment by zejji** [comment]
@ajcvickers @bricelam - I recently worked on a microservice-based project that depended on making EF Core migrations production-ready and thought it might be worthwhile sharing my experience.

At a very high level, our aim was to have each microservice be capable of initializing itself and its infrastructure (i.e. the [autopilot pattern](https://www.joyent.com/blog/app-centric-micro-orchestration)), so that deployment was as simple as starting one or more microservice containers. This massively simplified the deployment process.

More specifically, we needed each service to be able to:

- run a set of initialization tasks on application startup (including, but not limited to, database migrations. Registration of Kafka Connect connectors was another required task);
- ensure that only one instance of the app could run such tasks simultaneously; and
- use dependency injection in database migrations. This was required because, when running the migrations, we also needed to move data from legacy systems as well as simply make schema changes.

The approach we settled on had the following core features.

1. The infrastructure initialization feature could be added using a simple `IHost` extension method:
``` cs
public static void Main(string[] args)
{
   CreateHostBuilder(args).Build().RunUsingInfrastructureInitializer();
}
```
Configuration could also be used to either (i) bypass the infrastructure initializer; or (ii) only run the infrastructure initializer, but not boot the web application.

2. Under the hood, an `IInfrastructureInitializer` abstraction was responsible for running one or more IInfrastructureInitializationTasks. The  `IInfrastructureInitializer` used an `IDistributedLockProvider` (see e.g. [this project](https://github.com/madelson/DistributedLock)) to prevent parallel execution of tasks by multiple application instances. The interface was as follows:
``` cs
public interface IInfrastructureInitializer
{
    public void RunTasks();
}
```

3. An `IInfrastructureInitializationTask` abstraction was used for the tasks themselves. We implemented several common tasks, such as `DatabaseInitializationTask` (with allowed the target migration to be configurable) and `ConnectorRegistrationTask` and added extension methods to register these tasks with the DI container, but developers also had the flexibility to implement and register any other tasks desired.
``` cs
public interface IInfrastructureInitializationTask
{
    public void Run();
}
```

4. The final part of the jigsaw was to allow dependency injection to be used in migrations. We needed this specifically in order to move data from a legacy system at the same time as the migrations.

However, in my view, migrations can be useful as a broader tool to ensure that **any** task is performed only once in the entire history of an application and I can see no good reason to limit their use to only making database schema migration changes! I know this does not play nicely with converting migrations to SQL scripts (which was not a requirement for us), but I think the benefits of having the flexibility to use migrations in this way far outweighs the downsides.
 
Permitting the use of DI in migrations required a custom `MigrationsAssembly` as follows:
``` cs
#pragma warning disable EF1001 // Internal EF Core API usage.
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Diagnostics;
using Microsoft.EntityFrameworkCore.Infrastructure;
using Microsoft.EntityFrameworkCore.Migrations;
using Microsoft.EntityFrameworkCore.Migrations.Internal;
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Linq;
using System.Reflection;

namespace MyCompany.InfrastructureInitialization.Database
{
    /// <summary>
    /// A custom MigrationsAssembly which is modified to allow dependency injection
    /// to be used in migrations.
    /// </summary>
    public class DependencyInjectionEnabledMigrationsAssembly : MigrationsAssembly
    {
        private readonly IServiceProvider _serviceProvider;

        public DependencyInjectionEnabledMigrationsAssembly(
            ICurrentDbContext currentContext,
            IDbContextOptions options,
            IMigrationsIdGenerator idGenerator,
            IDiagnosticsLogger<DbLoggerCategory.Migrations> logger,
            IServiceProvider serviceProvider)
            : base(currentContext, options, idGenerator, logger)
        {
            _serviceProvider = serviceProvider;
        }

        /// <summary>
        /// This method is responsible for creating instances of the migration classes.
        /// </summary>
        /// <param name="migrationClass">The TypeInfo for the migration class to create</param>
        /// <param name="activeProvider">The name of the current database provider e.g. Microsoft.EntityFrameworkCore.SqlServer</param>
        /// <returns>A migration</returns>
        public override Migration CreateMigration(TypeInfo migrationClass, string activeProvider)
        {
            if (activeProvider == null)
            {
                throw new ArgumentNullException(nameof(activeProvider));
            }

            var constructors = migrationClass.GetConstructors();
            var hasConstructorWithParameters = constructors.Any(ci => ci.GetParameters().Length > 0);

            if (hasConstructorWithParameters)
            {
                var instance = ActivatorUtilities.CreateInstance(_serviceProvider, migrationClass.AsType());

                if (instance == null)
                {
                    throw new InvalidOperationException($"Could not create migration of type {migrationClass.Name}");
                }

                var migration = (Migration)instance;
                migration.ActiveProvider = activeProvider;
                return migration;
            }

            return base.CreateMigration(migrationClass, activeProvider);
        }
    }
}
#pragma warning restore EF1001 // Internal EF Core API usage.
```

Developers could then use any desired dependency within a migration by modifying the call to `.UseSqlServer()` as follows:

``` cs
optionsBuilder
    .UseSqlServer(configuration.GetConnectionString("testDatabase"))
    .UseInternalServiceProvider(new ServiceCollection()
        .AddEntityFrameworkSqlServer()
        .AddScoped<IMigrationsAssembly, DependencyInjectionEnabledMigrationsAssembly>()
        // add any additional dependencies you wish to inject into migrations here, for example:
        .AddSingleton<IDependency, Dependency>()
        .BuildServiceProvider());
```

While I am sure there other ways of achieving our goals, the above solution actually worked nicely for our purposes and involved a relatively small number of high-level abstractions which were easy to work with.

Happy to answer any questions if the above is of any interest.

**Comment by joaopgrassi** [comment]
@zejji whoa this looks really interesting! thank you for sharing it!

**Comment by roji** [comment]
> I know this does not play nicely with converting migrations to SQL scripts (which was not a requirement for us), but I think the benefits of having the flexibility to use migrations in this way far outweighs the downsides.

My general recommendation would be to avoid *any* sort of automated application of migrations to production database without previously confirming that the actual SQL to be executed is what you want; EF does its best to produce correct migration SQL, but in some cases scripts need to be corrected, and blindly applying them could result in pretty serious data loss (the typical example is an intended column rename which causes the old column to be dropped and a new one to be created).

in most scenarios, this means manual inspection of the SQL scripts - unless you invest in automated application of migrations to a test database already containing data, and thorough automated testing of your application against that database (this is non-trivial).

**Comment by zejji** [comment]
@roji - Yes, it probably goes without saying that the process described in my post above depends on going through several stages of automated and manual before going anywhere near production.

**Comment by JonPSmith** [comment]
@zejji. I really like your approach. Can you provide a link to the code that ensures only one instance of a multi-instance web service would apply a migration.

**Comment by zejji** [comment]
@JonPSmith - The basic idea of the `InfrastructureInitializer` class itself is pretty simple, leveraging a custom wrapper round the [DistributedLock](https://github.com/madelson/DistributedLock) library which just added the ability to auto-create the lock database. As mentioned previously, however, there is an argument for using an explicit task registration method rather than piggy-backing off the DI container:

``` cs
public class InfrastructureInitializer : IInfrastructureInitializer
{
    private const string GlobalLockName = "InitializationLock";
    private const int DefaultLockTimeoutInSeconds = 120;

    private readonly IDistributedLockProvider _synchronizationProvider;
    private readonly IEnumerable<IInfrastructureInitializationTask> _initializationTasks;
    private readonly int _lockTimeoutInSeconds;

    public InfrastructureInitializer(
        IDistributedLockProvider synchronizationProvider,
        IEnumerable<IInfrastructureInitializationTask> initializationTasks,
        IOptions<InfrastructureInitializerSettings> options)
    {
        _synchronizationProvider = synchronizationProvider;
        _initializationTasks = initializationTasks;
        _lockTimeoutInSeconds = options?.Value?.AcquireLockTimeoutInSeconds ?? DefaultLockTimeoutInSeconds;
    }

    public void RunTasks()
    {
        var globalLock = _synchronizationProvider.CreateLock(GlobalLockName);
        using (globalLock.Acquire(TimeSpan.FromSeconds(_lockTimeoutInSeconds)))
        {
            foreach (var initializationTask in _initializationTasks)
            {
                initializationTask.Run();
            }
        }
    }
}
```

The tests proving that it works were naturally much more involved than the code itself, as they involved e.g. (i) verifying the lock worked by creating multiple service providers and triggering the initialization jobs in a Parallel.ForEach; and (ii) ensuring that initialization tasks were executed in the write order.

The database initialization task was as follows:

``` cs
public class DatabaseInitializationTask : IInfrastructureInitializationTask
{
    private readonly IMigrationDbContext _context;
    private readonly string? _targetMigration;

    public DatabaseInitializationTask(
        IMigrationDbContext context,
        IOptions<DatabaseInitializationTaskSettings> options)
    {
        _context = context;

        _targetMigration = options?.Value?.TargetMigration;
        if (string.IsNullOrWhiteSpace(_targetMigration))
        {
            _targetMigration = null;
        }
    }

    public void Run()
    {
        // Cf. https://blog.rsuter.com/automatically-migrate-your-entity-framework-core-managed-database-on-asp-net-core-application-start/
        var migrator = _context.Database
            .GetInfrastructure()
            .GetRequiredService<IMigrator>();

        if (migrator != null)
        {
            // See https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.migrations.imigrator.migrate?view=efcore-5.0
            // All migrations will be run if targetMigration is null.
            // If targetMigration is a string (e.g. "2021xxxxxxxxx_InitialCreate") to migrate to a specific target.
            migrator.Migrate(_targetMigration);
        }
    }
}
```

**Comment by roji** [comment]
@zejji reading what I wrote, my intention was indeed more to stress the importance of inspecting before applying - whether the SQL script is inspected or the migration is indeed less important; I should have been more clear on that point. We get lots of issues from people who seem to apply migrations blindly (and upon program startup), so this point is always worth stressing.

I do still believe inspecting SQL is valuable in this context, since migration code doesn't always translate to SQL in a completely transparent way; but a proper testing infrastructure and deployment process can probably be sufficient.

> And if you can't trust migrations which have been tested on the basis that they are not raw SQL, should you never use an ORM to manipulate data either?

It's not quite a matter of trust - synchronization schema changes contain some inherent difficulties which don't exist in data manipulation (the column renaming is a good example). It's simply a riskier thing.

**Comment by zejji** [comment]
@roji - Agree with everything you said and perhaps I was being a little flippant. Column renaming and similar gotchas certainly exist but, from experience, provided you do not assume auto-generated migrations can be trusted, this risk is entirely manageable.

**Comment by ajcvickers** [comment]
@zejji This is all really interesting; we should see what we can do about getting this kind of approach added to our guidance. Also, would you be interested in coming on the [EF Core Community Standup](https://dotnet.microsoft.com/live/community-standup) to share this with others?

/cc @JeremyLikness @bricelam and also @glennc since this is very microservices related.

**Comment by zejji** [comment]
@ajcvickers - I'd be very happy to share on the EF Core Community standup if it's useful to people; just let me know what's needed and I'll prepare something accordingly.

**Comment by cResults** [comment]
not sure whether or not this is the correct place for this feedback, but being able to more easily apply EF Migrations to all database instances associated with a DbContext within Elastic Database shards would be a huge benefit to developers using the database per tenant modal of multi-tenancy.  We're currently working to write a process to run via Azure DevOps pipeline that will use ef cli connected to one instance of our dbs to create the diff script and then use powershell to create an Elastic Job to apply that diff script to all dbs. 

We're still primarily EF 6.x, but we have a fair number of new services that are ef core.

If there is an easier way to apply updates to 100's of the same database in a sharded environment, please point us to it.  If an easier way doesn't exist and you'd like to build it let us know, we may be willing to help.

**Comment by ajcvickers** [comment]
@zejji If you haven't already and you are still interested, then could you go to https://aka.ms/efstanduprequest so we can get this show on the schedule. /cc @JeremyLikness

**Comment by nick5454** [comment]
I created a Cake package the simply calls EF's command line to get a list of all migrations in the project and then create a script file for every migration. On the release side, I then deploy a small powershell script that checks for the migration and if not exists we deploy the script. This way no matter what server we deploy to, ef will update to be current.

Works really well without needed to have an EF live database connection.
`Debug("Starting Migration");
		if(!string.IsNullOrWhiteSpace(BuildParameters.DbContext) &&
			!string.IsNullOrWhiteSpace(BuildParameters.StartupProject) &&
			!string.IsNullOrWhiteSpace(BuildParameters.DomainProject))
		{
			string zipDirectory = $"{BuildParameters.Artifacts}\\zip\\";

			//string efOutput = $"{zipDirectory}\\{domainOutputSql}";
			int efCount = 0;
			string result = "";

			while( string.IsNullOrWhiteSpace(result) && efCount < 5) {
				var efSettings = new DotNetCoreEfMigrationListerSettings() {
					StartupProject = GetRelativePathToProject(BuildParameters.StartupProject).ToString(),
					Project = GetRelativePathToProject(BuildParameters.DomainProject).ToString(),
					Configuration = BuildParameters.Configuration,
					NoBuild = true,
					Verbose = true

				};
				Debug($"Trying EF Migration list time {efCount}");

				Debug($"Start EF call");
				Debug($"Project Collection: {GetRelativePathToProject(BuildParameters.DomainProject).ToString()}");
				result = await DotNetCoreEfMigrationList(GetRelativePathToProject(BuildParameters.DomainProject).ToString(), efSettings);
				Debug($"result = {result}");
				Debug($"End EF Call");

				efCount++;

			}
			Debug($"result = {result}");

			if(!string.IsNullOrWhiteSpace(result)) {
				//DotNetCoreEfMigrationScript("/", efSettings);
				var migrations = Newtonsoft.Json.JsonConvert.DeserializeObject<List<MigrateValueObject>>(result);
			
				Debug("Looping migration");
				EnsureDirectoryExists($"{zipDirectory}SQL");

				string lastMigrate = "0";

				foreach(MigrateValueObject migrate in migrations) {
	
					try {			
						string efOutput = $"{zipDirectory}SQL\\{migrate.id}.sql";

						var efMigrateSettings = new DotNetCoreEfMigrationScriptSettings() {
							StartupProject = GetRelativePathToProject(BuildParameters.StartupProject).ToString(),
							Context = BuildParameters.DbContext,
							Project = GetRelativePathToProject(BuildParameters.DomainProject).ToString(),
							Configuration = BuildParameters.Configuration,
							Output = efOutput,
							Idempotent = true,
							NoBuild = true,
							From = lastMigrate,
							To = migrate.id
							};

						DotNetCoreEfMigrationScript("/", efMigrateSettings);
					} catch ( Exception ex ) {
						Debug($"Error '{migrate.id} Exception: {ex}");
					}

					lastMigrate = migrate.id;
				}
			}

			//DotNetCoreEfDatabaseUpdate("/", efSettings);
		} else {
			Debug("EF BUILD: Not all values exist. Skipping... ");
		}

		Debug("migration complete");`

and the powershell `param ($connString, $sqlFolder)

  Write-Host "Connstring=$($connString)"
  Write-Host "Folder=$($sqlFolder)"

try {
    
   # Invoke-Sqlcmd -ConnectionString

  $query = "Select * from Users"

 # Invoke-Sqlcmd -ConnectionString $connString -Query $query

  $table = get-childitem -recurse | where {! $sqlFolder}
  $table = Get-ChildItem -Path $sqlFolder -File 
  $exists = "false"

  $timeout = New-Object System.TimeSpan -ArgumentList @(5,0,0) # five hours
  $options = [System.Transactions.TransactionScopeOption]::Required


  # all scripts in 1 transaction. If any fail then rollback ALL of THEM
 

  foreach($row in $table) 
  {
     $file = (Split-Path $row.FullName -Leaf).Split('.')[0]
     Write-Host $file

     Write-Host "Determining if we should fail migration"

    Write-Host $connString

     $scope = New-Object -TypeName System.Transactions.TransactionScope -ArgumentList @($options, $timeout)

    # Invoke-Sqlcmd -ConnectionString $connString -InputFile $row.FullName
    $DS = Invoke-Sqlcmd -ConnectionString $connString -Query "SELECT Count(*) as Total FROM [__EFMigrationsHistory] WHERE MigrationId = '$($file)'"
    
    if( $DS.Total -eq 0 ) {
        $exists = "false"
    } else {
        $exists = "true"
    }
    
    Write-Host $exists

    # if exists then bypass migration
    if( $exists -eq "false") {
        Invoke-Sqlcmd -ConnectionString $connString -InputFile $row.FullName -ea stop 
        
    } 

    $scope.Complete()
    $scope.Dispose() 
  }

  
} catch [System.Net.WebException],[System.IO.IOException] {
  Write-Host "Errors"
} catch {
  Write-Host "Error executing script"
  Write-Host $_

  if ($exists -eq "false") 
  { 
    Write-Host "Error with new migration. Exiting..."
    break 
  }
} finally {
    $scope.Dispose() 
}

`

**Comment by JonPSmith** [comment]
Hi @zejji,

Your solution to ensure only one application migrates the database allowed me to (finally) implement an open-source library that provides [another way to handle authorization](https://www.thereformedprogrammer.net/a-better-way-to-handle-authorization-in-asp-net-core/) in ASP.NET Core. 

But for this library to work I need cover all of the ASP.NET Core startup scenarios, including multiple applications happening at the same time. When you [described your approach](https://github.com/dotnet/efcore/issues/19587#issuecomment-796693956) you said:
>... we ended up using a poll loop and appropriate error handling and it worked acceptably even in the face of multiple instances being fired-up simultaneously for test purposes.

I wonder if you could provide some more information on what the "poll loop and appropriate error handling" needs to do, as checking these race states are difficult to create and test.

**Comment by zejji** [comment]
@JonPSmith - Since writing the above post, I have moved to a new job and unfortunately no longer have access to the original code outlined above.

If I recall correctly, the approach I took was to use EF Core's internal [IRelationalDatabaseCreator](https://github.com/dotnet/efcore/blob/554043340b91299aa7e887fc69efe312a6720732/src/EFCore.Relational/Storage/IRelationalDatabaseCreator.cs), e.g. the [SqlServerDatabaseCreator](https://github.com/dotnet/efcore/blob/554043340b91299aa7e887fc69efe312a6720732/src/EFCore.SqlServer/Storage/Internal/SqlServerDatabaseCreator.cs), to create the database if it didn't exist. (This is used by EF's [Migrator](https://github.com/dotnet/efcore/blob/554043340b91299aa7e887fc69efe312a6720732/src/EFCore.Relational/Migrations/Internal/Migrator.cs) class. However, since it is an internal API the downside is that it is potentially subject to change.)

Obviously it's not possible to create a lock at that stage, but I established through testing that (i) an exception will always be thrown by EF - I can't recall whether it was a SqlException or something more specific - when using SQL Server if two services attempt to create the same database at the same time, (ii) this exception can be handled gracefully; and (iii) the database was nonetheless always created in a consistent state despite the race condition.

So I took the approach of running the following steps in a loop, with a max number of retries:

 - checking whether the database existed;
 - if not, attempting to create the database;
 - if an exception was thrown, waiting a short time before trying again.

Obviously the database creation step only needs to happen once, so this approach proved to be suitable for our purposes.

**Comment by JonPSmith** [comment]
Hi @zejji,

Thanks for getting back to me, and your explanation is a great help. I'll work on this in my library and once its done I'll let you know. 

PS. I managed to add a link to your comment on this approach to chapter on EF Core migrations my book, [Entity Framework Core in Action, 2nd edition](https://bit.ly/EfCoreBookEd2) just before it went to print. If you contact me via my [web site](https://www.thereformedprogrammer.net/contact/) I'll see if I can get you a free ebook.

**Comment by dazinator** [comment]
@zejji @JonPSmith 
With respect to the distributed locks database, one option is to use a file based distributed lock implementation instead, and configure a central file share as the locks directory. This should be enough to establish a lock to create any SQL databases and run the infrastructure tasks. Creating of the a SQL server based locks database could then be one of these tasks running within the file based lock supposing you really wanted to use SQL server database for distributed locks within your application from that point onwards. Your IDistributedLockProvider could be registered with a factory that returns a file based implementation when the infrastructure tasks are incompleted, and returns a SQL based lock provider when the setup tasks are complete. Bit more complicated but might be more suitable if you already have a file share in place.

**Comment by JonPSmith** [comment]
@dazinator @zejji,

Thanks @dazinator  for taking the time to look at this. I like it, but for a library which could be used in monoliths, microservices, serverless, containers etc. I can't be sure there is a common file location. I need to think a bit more on this, but do add anything if you think it would help.

I'm also thinking of building a library just for EF Core create/migrate usage, as I think others would like it too. @zejji, I don't want to take over your idea, so if you want to do something yourself then I'm happy to not do that. @bricelam, @ajcvickers I assume the EF Core team isn't planning to pick this up??

**Comment by zejji** [comment]
@JonPSmith - Thanks for your kind words and glad to hear it's been helpful. 😃

@dazinator - Yes, that's definitely a good approach if some other locking mechanism is already available prior to the database creation (e.g. [Azure blob leases](https://github.com/madelson/DistributedLock/blob/master/docs/DistributedLock.Azure.md)). The most flexible approach would, as you say, be to work in a configuration option so an alternative locking mechanism could be used for the initial database creation, falling back to the default of a poll loop.

**Comment by zejji** [comment]
@JonPSmith - I agree that it would be useful to put the logic in a publicly-available library, as I think it could be useful wherever people are running migrations in a service with potentially more than one replica (and want to avoid complex deployment scripts). I don't have access to the original source code anymore, so would need to recreate it. However, I'm not precious about ownership so if you are already well-advanced I certainly wouldn't object to your using the idea 😃

**Comment by JonPSmith** [comment]
@zejji, I haven't started on it yet, but was getting input for when I do have to build it. I'm busy with the library that needs this feature so if you have the time/inclination to build such a library, then by all means do that. I am happy to collaborate with you, but we should wait for feedback from @bricelam, @ajcvickers in case the EF Core team want to pick this up.

Also @zejji, @dazinator. I agree that providing a way for the user to use one of the other [DistributedLock](https://github.com/madelson/DistributedLock) approaches should give people enough options to get around this.

One thing I have been thinking about is your `InitializationTask` concept which allows other, one-off tasks that need to be executed within the lock - for instance my library would want to seed the database using setup data in the appsettings.json after the migration was finished. These `InitializationTasks` could be run in the order that they were registered with the DI provider, but a more secure way would be to make the `IInitializationTask` have a `public Type RunThisAfterTask = typeof(<task that should precede this task>)` property to ensure the tasks are run in the right order.

**Comment by roji** [comment]
> I am happy to collaborate with you, but we should wait for feedback from @bricelam, @ajcvickers in case the EF Core team want to pick this up.

We generally recommend applying migrations as part of the application deployment process, and not as part of application startup. Among other things, this side-steps the whole distributed locking question, since that step is being performed at one place and not concurrently. Even if you have a distributed locking mechanism, you must still likely ensure that application instances are down or waiting for the migration to be applied, otherwise you get different versions of your application trying to access a different database schema... And that begs the question of why do it as part of application startup in the first place...

Even if you want to use distributed locking, the specific mechanisms for doing so are very different across databases (and don't necessary exist on all of them). I'm not sure what EF Core should or could contribute here.

**Comment by gezhl** [comment]
> Even if you have a distributed locking mechanism, you must still likely ensure that application instances are down or waiting for the migration to be applied, otherwise you get different versions of your application trying to access a different database schema...

@roji - The approach we took was that zero-downtime deployments were fine as long as the [expand-contract pattern](https://medium.com/continuousdelivery/expand-contract-pattern-and-continuous-delivery-of-databases-4cfa00c23d2e) was used, whereas any other migrations would require instances to be brought down. Since, in practice, most migrations tend to fall within the former category, I still think there's a strong argument for doing the changes at application startup.

**Comment by zejji** [comment]
> Even if you have a distributed locking mechanism, you must still likely ensure that application instances are down or waiting for the migration to be applied, otherwise you get different versions of your application trying to access a different database schema...

@roji - The approach we took was that zero-downtime deployments were fine as long as the expand-contract pattern was used, whereas any other migrations would require instances to be brought down. Since, in practice, most migrations tend to fall within the former category, I still think there's a strong argument for doing the changes at application startup.

**Comment by JonPSmith** [comment]
hi @roji,

Thanks for clarifying that - I should have remembered the EF Core philosophy for migrations. And yes, the DistributedLock library only supports SQL Server and Postgresql databases, which wouldn't match the EF Core team's aims. 

@zejji and/or me will look at this because I still think @zejji approach is really useful. As well as making migrations possible on startup it also allows an action that has to be run once on startup - in ASP.NET Core authorization library I am building there is a need to seed the database with certain settings provided by the appsettings.json file, which most be only done once.

**Comment by roji** [comment]
> @roji - The approach we took was that zero-downtime deployments were fine as long as the expand-contract pattern was used, whereas any other migrations would require instances to be brought down. Since, in practice, most migrations tend to fall within the former category, I still think there's a strong argument for doing the changes at application startup.

That's a interesting point - it's indeed a good idea to manage migrations in a way which doesn't require downtime. But then I'm still unclear on the concrete advantages of applying at startup as opposed to at deployment... At the very least, the latter saves you from having to deal with the distributed locking issue ([more discussion is available in our docs](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/applying?tabs=dotnet-core-cli)).

**Comment by zejji** [comment]
@roji - Running the migrations at application startup wasn't actually our first choice and our original intention was to have one or more deployment scripts which handled all the infrastructure initialization tasks.

However, because our application was distributed to clients for deployment, this initialization logic couldn't be kept in, e.g. Azure DevOps pipelines running for each service, and was instead initially encapsulated in a single Ansible deployment script. Another possible option might have been to use Kubernetes init containers. In either case, we found that the complexity of the infrastructure-as-code was spiraling out of control. This made it difficult to understand for our team.

Having a simple library (as described above) which was used in each microservice made things much easier. Since all initialization logic was now encapsulated in each service, our dev docker-compose file could be as simple as just a list of the services themselves. In other words, this now represented the entirety of the knowledge required to deploy the application.

This may be less relevant for a monolithic application but, in any case, there is certainly some appeal for me in having an application able to safely initialize its infrastructure in a few lines of code.

**Comment by roji** [comment]
Thanks for the additional details @zejji.

**Comment by joaopgrassi** [comment]
Just to say something, not sure it's going to help anybody, definitely not Jon and the idea of a library but: We started rolling the migrations at startup as well in our k8s cluster. We were "lucky" that we have in our infrastructure always one single pod that did some background work, so we just made that pod run the migrations during startup on each new deployment.

But, we started to have cases where we didn't have this "single" pod anymore and we were at this point where couldn't run the migrations as part of the service's startup since we have many instances of it in k8s. 

The approach I used and it's working so far for us is use k8s jobs + helm pre-install hooks. Basically the idea is that a service declares a [pre-install Helm hook](https://helm.sh/docs/topics/charts_hooks/#writing-a-hook) and via a "simple" console app, we run the migrations and all happens before Helm upgrades the charts.

Like @roji said, this also has downsides as currently running pods will use a new version of the db for a fraction of time, but in our team so far we are being diligent and managed to not introduce any backwards incompatible changes, so all works. 

Another point which we didn't do yet is to handle cases where the migration fails.. but AFAIK we could also leverage hooks for that, and rollback it using  maybe `post-rollback` hooks.

**Comment by dazinator** [comment]
> > Even if you have a distributed locking mechanism, you must still likely ensure that application instances are down or waiting for the migration to be applied, otherwise you get different versions of your application trying to access a different database schema...
> 
> @roji - The approach we took was that zero-downtime deployments were fine as long as the expand-contract pattern was used, whereas any other migrations would require instances to be brought down. Since, in practice, most migrations tend to fall within the former category, I still think there's a strong argument for doing the changes at application startup.

I've been using the expand contract principle for years but never knew that was it's formal term! Thank you for this - now I can sound more educated when talking to my fellow developers :-)
On a related note, I wonder if there is anything EF can do to highlight:

1. Potentially breaking migrations. For example - adding a new non nullable column with no default value is potentially breaking when there are existing records. 
2. Aid with expand contract pattern. For example, removing a column is a contraction that may break existing instances and is therefore more "unsafe" than an expansion of the model. Imagine if by default "contractions" could be prevented by EF migrations unless you pass in a special "-allow-contraction" argument when generating the model. This would make it safer to use the expand contract pattern without accidental premature contractions. P.s "accidental premature contractions" is not a phrase I'd ever imagine writing in relation to software dev :-)

**Comment by roji** [comment]
> I've been using the expand contract principle for years but never knew that was it's formal term! Thank you for this - now I can sound more educated when talking to my fellow developers :-)

Oh I just made up the term zero-downtime migrations, definitely nothing formal about it :rofl: 

> adding a new non nullable column with no default value is potentially breaking when there are existing records.

When you do this with EF Core, it will create a non-nullable column with the the default value for existing columns (e.g. 0 for ints - but you can go change the default in the scaffolded migration if you want). So this isn't a breaking migration.

> For example, removing a column is a contraction that may break existing instances and is therefore more "unsafe" than an expansion of the model

EF does already issue a warning about potentially destructive migrations (e.g. removing a column). That sounds like what you're describing here?

**Comment by dazinator** [comment]
> Oh I just made up the term zero-downtime migrations, definitely nothing formal about it 

I was referring to the term "expand-contract pattern" mentioned by @zejji :-) it's nice to know this formal pattern name at last!

> EF does already issue a warning about potentially destructive migrations (e.g. removing a column). That sounds like what you're describing here?

Yes but I think to support the expand-contraxt pattern more formally, you'd want ef core migration generation to error for any contraction of the model. This could be removal of a column but it could also be a reduction in column size or the change to a datatype of a column etc - basically anything that isn't purely additive / an expansion of the model can be viewed as dangerous and so it would be great for safety to reasons to have ef migration generation "fail by default" in such a case as opposed to just generating warnings which can easily be missed. Then for times where you don't want this safety net- i.e because you want to do a legitimate "contraction" of the model, you could perhaps generate the EF migration with an optional command line parameter that turns this safety net off and allows the migration to be created. Just an idea - but off topic so I apologise:-)

**Comment by roji** [comment]
@dazinator it sounds like you're asking for a way to turn the destructive migration warning into an error?

**Comment by dazinator** [comment]
@roji yes if we could make `dotnet ef migration add` error by default for destructive migrations that would be safer for us. With a way to override when needed, for example suppose we pass in an additional command line arg, something like `--allow-destructive`. Finally if the destructive migrations themselves were generated with an informational attribute added to the top of the migration class, they could be more easily seen and discussed during code / peer review phase

```csharp
[Destructive()]
public class SomeMigration
{

}
```

**Comment by JonPSmith** [comment]
Hi @zejji and @ajcvickers,

I have finally created a library that uses @zejji [DistributedLock](https://github.com/madelson/DistributedLock) approach. You can see the code at [RunStartupMethodsSequentially](https://github.com/JonPSmith/RunStartupMethodsSequentially) and I have also written an article called ["How to safely apply an EF Core migrate on ASP.NET Core startup"](https://bit.ly/SafeMigrateArticle) which has a lot of info in it.

@zejji, you will see I overcame the "database not created yet" by having two parts to the locking:

1. Check if the resource exists. If doesn’t exist, then try another lock approach
 2. If the resource exists, then lock that resource and run the startup services

This means I can try to lock the database, but if its not there i have a second go using a FileStore Directory. 
I have tried it on Azure and it works (see the article). @zejji, I really appreciate you describing this approach.

@ajcvickers  and @roji: I know you don't like the "migrate on startup" approach but I have made it very clear that you need to make sure your migration does not contain a _breaking change_ migration.

**Comment by zejji** [comment]
@JonPSmith - Thanks for the heads-up regarding your library implementation. Looks good! 😃

**Comment by GeorgeTTD** [comment]
> ```cs
> [Destructive()]
> public partial class SomeMigration
> {
> 
> }
> ```



> If there are no destructive migrations however then you might do a zero downtime deployment and care less about the above.
> 
> ```
> dotnet ef database upgrade -connection [conn string] --has-destructive
> True
> ```

@dazinator @roji I would like to add another angle to this. We are currently looking at custom implementations to do a zero downtime deploy for apps with EF. Our prefered deploy strategy would be:

1. Spin up new deployment 2
2. Apply non destructive migrations
3. Switch traffic to route to deployment 2
4. remove deployment 1
5. Apply destructive migrations

It would be nice if you could split Up and Down migrations into destructive and non-destructive. You could do this several ways so I will leave the implementation up to the powers that be. However here is an idea which would work for our scenarios.

```
dotnet ef migrations add MyMigration --split-destructive

Outputs
# <timestamp>_MyMigration.cs
# <timestamp>_destructive_MyMigration.cs


dotnet ef database upgrde ...

# Only applies <timestamp>_MyMigration.cs

and then once deploy is done and deployment 1 is remove

dotnet ef database upgrde ... --allow-destructive

# applies <timestamp>_destructive_MyMigration.cs
```

**Comment by roji** [comment]
@GeorgeTTD how would the proposed --split-destructive switch work if there are more than one migrations pending, each with its own non-destructive and destructive components?

More generally, you are of course free to structure your migrations in this way, and that's indeed what zero-downtime migrations usually entail; but this isn't really something EF Core can automate for you without your own planning and inspection of migrations.

**Comment by dazinator** [comment]
@GeorgeTTD @roji If EF did mark destructive migrations with an attribute e.g 
```
[Destructive()]
public partial class SomeMigration
{

}
```
Then @GeorgeTTD you could atleast get part way there with the following:-
1. Use two seperate migrations assemblies, for destructive vs non destructive.
2. Add a test to assert this convention - uses reflection to discover migrations  with or without this attribute and makes sure they haven't accidentally been added to the wrong project.
3. You can now run each migrations assembly independently as required..

However I'm not sure this makes sense as could you get scenarios where a non destructive migration depends on a destructive one if that was the order they were generated. For example a migration to do a column rename (destructive) could be generated before a migration to add an index on that column (non destructive).. For this reason it might be the best you can do is just to know whether the release does or doesn't contain non destructive migrations and optimise your deployment accordingly.

**Comment by dazinator** [comment]
Just to be clear, my proposition is for EF to support a workflow that prevents teams from accidentally including Destructive migrations in a release. The value is that by deterring this it allows deployment flows where services need not be brought down whilst the database is being upgraded. It allows support of workflows where releases containing destructive migrations can be done, but behind a safety net that enforces ream communication and planning.

**Comment by roji** [comment]
@dazinator as you write above, migrations are a linear, ordered list, with dependencies between them. It isn't possible to execute only non-destructive migrations, deferring destructive ones that are interspersed between them (because of the dependencies). So splitting migrations into separate assemblies (or otherwise cherry-pick/separating them) doesn't seem possible.

However, in your second comment you seem to be proposing a cmdline switch where the EF Core tooling would simply refuse to apply migrations (e.g. when running a migrations bundle or `dotnet ef database update`) if any of them contain (or come after) a destructive migration. That does seem possible, and could possibly help prevent accidental errors, though note that any migrating would be blocked once a destructive migration is pending; I'm not sure exactly how that helps the workflow.

At the end of the day, zero-downtime migrations are something which require careful planning and a specific workflow; EF can certainly try to help, but nothing will obviate a manual checking and migration planning here.

Regardless, if you want to continue discussing this, please open a new issue - let's try to keep this general issue clear of detailed specific discussions.

**Comment by stijnherreman** [comment]
> In our latest project we integrated the DB migrations in the deployment pipeline simply by generating an idempotent migration script (for all migration steps) via dotnet-ef and leveraging sqlcmd to execute it against the appropriate sql server instance.

I ended up with the same solution. The idempotent script is an artifact published by the build stage and is executed with sqlcmd.

```yaml
- task: PowerShell@2
  inputs:
    targetType: inline
    script: |
      Import-Module SqlServer
      Invoke-Sqlcmd -ServerInstance <...> -Database <...> -InputFile <...>
```

One thing I'd still like to add is the ability to review the generated script in a pull request. GitHub [has done this](https://github.blog/2020-02-14-automating-mysql-schema-migrations-with-github-actions-and-more/) and open sourced their code, so I'm hoping to use their work as inspiration.

**Comment by ajcvickers** [comment]
@alrz If you want to write a migration where you explicitly call the builder methods instead of scaffolding the starting point from the model, then you are free to do that. You can then create a script from this in the normal way.
